lib/path/default.nix

```diff
@@ -271,6 +271,51 @@ in /* No rec! Add dependencies on this file at the top. */ {
               second argument: "${toString path2}" with root "${toString path2Deconstructed.root}"'';
         joinRelPath components;
 
+  /*
+  Split the filesystem root from a path.
+  The result is an attribute set with these attributes:
+
+  - `root`: The filesystem root of the path, meaning that this directory has no parent directory.
+  - `subpath`: The normalised subpath string (see `lib.path.subpath.normalise) that when appended (see `lib.path.append`) to `root` returns the original path.
+
+  Laws:
+
+  - Appending the `root` and `subpath` gives the original path:
+
+        p ==
+          append
+            (splitRoot p).root
+            (splitRoot p).subpath
+
+  - Trying to get the parent directory of `root` using `readDir` returns `root` itself:
+
+        dirOf (splitRoot p).root == (splitRoot p).root
+
+  Type:
+    splitRoot :: Path -> { root :: Path, subpath :: String }
+
+  Example:
+    splitRoot /foo/bar
+    => { root = /.; subpath = "./foo/bar"; }
+    splitRoot /.
+    => { root = /.; subpath = "./."; }
+    # Nix neutralises `..` path components for all paths automatically
+    splitRoot /foo/../bar
+    => { root = /.; subpath = "./bar"; }
+    splitRoot "/foo/bar"
+    => <error>
+  */
+  splitRoot = path:
+    assert assertMsg
+      (isPath path)
+      "lib.path.splitRoot: Argument is of type ${typeOf path}, but a path was expected";
+    let
+      deconstructed = deconstructPath path;
+    in {
+      root = deconstructed.root;
+      subpath = joinRelPath deconstructed.components;
+    };
```

Comment:
This does look a bit redundant. I remember we had some of that discussion, but why exactly do we normalise back the path instead of just exposing `deconstructPath` as `splitRoot`?

Reply:
In https://github.com/NixOS/nixpkgs/pull/238013 we agreed to make `removePrefix` return a subpath. It should also be subpath here for similar reasons. It fits nicely into the rest of the library, because we can just say that `append root subpath == path`. Otherwise this function would be the only oddball returning path components as a list.

I also like to think of [`lib.path.subpath.components`](https://github.com/NixOS/nixpkgs/pull/242695) being the only bridge out of the path/subpath world into the list world, where you can then use all of the `lib.lists` functions (such as [`commonPrefix`](https://github.com/NixOS/nixpkgs/pull/243520)), while `lib.path.subpath.join` is a way back from lists to subpaths (though maybe we should also have `lib.path.subpath.fromComponents` that does almost the same).

Reply:
Yeah right, that is consistent. We may want to record that decision in the [design document](https://github.com/NixOS/nixpkgs/blob/master/lib/path/README.md).

Reply:
I added this to the [`subpath.components` PR](https://github.com/NixOS/nixpkgs/pull/242695)